plugins/woocommerce/src/Internal/DataStores/Orders/FeatureController.php

```diff
@@ -0,0 +1,324 @@
+	 *
+	 * @param WC_Object_Data_Store_Interface|string $default_data_store The default data store (as received via the woocommerce_order_data_store hooks).
+	 * @return WC_Object_Data_Store_Interface|string The actual data store to use.
+	 */
+	private function get_data_store_instance( $default_data_store ) {
+		if ( $this->is_feature_visible() && $this->custom_orders_table_usage_is_enabled() && ! $this->data_synchronizer->data_regeneration_is_in_progress() ) {
+			return $this->data_store;
+		} else {
+			return $default_data_store;
+		}
+	}
+
+	/**
+	 * Add an entry to Status - Tools to create or regenerate the custom orders table,
+	 * and also an entry to delete the table as appropriate.
+	 *
+	 * @param array $tools_array The array of tools to add the tool to.
+	 * @return array The updated array of tools-
+	 */
+	private function add_initiate_regeneration_entry_to_tools_array(array $tools_array ): array {
+		if ( ! $this->is_feature_visible() ) {
+			return $tools_array;
+		}
+
+		$orders_table_exists       = $this->data_synchronizer->check_orders_table_exists();
+		$generation_is_in_progress = $this->data_synchronizer->data_regeneration_is_in_progress();
+
+		if ( $orders_table_exists ) {
+			$generate_item_name   = __( 'Regenerate the custom orders table', 'woocommerce' );
+			$generate_item_desc   = __( 'This tool will regenerate the custom orders table data from existing orders data from the posts table. This process may take a while.', 'woocommerce' );
+			$generate_item_return = __( 'Custom orders table is being regenerated', 'woocommerce' );
+			$generate_item_button = __( 'Regenerate', 'woocommerce' );
+		} else {
+			$generate_item_name   = __( 'Create and fill custom orders table', 'woocommerce' );
+			$generate_item_desc   = __( 'This tool will create the custom orders table and fill it with existing orders data from the posts table. This process may take a while.', 'woocommerce' );
+			$generate_item_return = __( 'Custom orders table is being filled', 'woocommerce' );
+			$generate_item_button = __( 'Create', 'woocommerce' );
+		}
+
+		$entry = array(
+			'name'             => $generate_item_name,
+			'desc'             => $generate_item_desc,
+			'requires_refresh' => true,
+			'callback'         => function() use ( $generate_item_return ) {
+				$this->initiate_regeneration_from_tools_page();
+				return $generate_item_return;
+			},
+		);
+
+		if ( $generation_is_in_progress ) {
+			$entry['button'] = sprintf(
+			/* translators: %d: How many orders have been processed so far. */
+				__( 'Filling in progress (%d)', 'woocommerce' ),
+				$this->data_synchronizer->get_regeneration_processed_orders_count()
+			);
+			$entry['disabled'] = true;
+		} else {
+			$entry['button'] = $generate_item_button;
+		}
+
+		$tools_array['regenerate_custom_orders_table'] = $entry;
+
+		if ( $orders_table_exists ) {
+
+			// Delete the table.
+
+			$tools_array['delete_custom_orders_table'] = array(
```

Comment:
these tools would probably change based on what we want in migration approach, but this is fine for now